### PR TITLE
Add a second endpoint with a single on/off cluster on it to the wifi-echo server

### LINKED ISSUE #2390: Need a second endpoint for the wifi-echo server
#### Problem
We want to demo multiple endpoints at some point.  Right now the wifi-echo server is configured with only one endpoint.

We used to have two endpoints, with the on/off cluster on both of them, before, when we had the dotdot generated code.

 #### Proposed Solution
Add second endpoint, maybe with just the on-off cluster on it at the moment.

I'm sorry I didn't catch this earlier...  Looking at how to do this most easily.

@rwalker-apple

### PR DESCRIPTION
**Pull request by bzbarsky-apple**
#### Problem
We want to control two independent lights, but the server is not configured for that.

 #### Summary of Changes
Add a second endpoint to represent the second light.

fixes https://github.com/project-chip/connectedhomeip/issues/2390

### CODE REVIEW

File: examples/wifi-echo/server/esp32/main/gen/endpoint_config.h [hunk 1]
```diff
@@ -262,12 +262,16 @@
                  CLUSTER_MASK_PRE_ATTRIBUTE_CHANGED_FUNCTION),                                                                     \
                 emberAfFuncArrayIasZoneClusterServer,                                                                              \
             },                                                                                                                     \
+            {                                                                                                                      \
+                0x0006, (EmberAfAttributeMetadata *) &(generatedAttributes[18]), 1,                                                \
```
Comment (rwalker-apple) [inline examples/wifi-echo/server/esp32/main/gen/endpoint_config.h:266]: this makes endpoint 2 point at the same boolean as endpoint 1.  You probably want generatedAttributes[84], instead

**Review by saurabhst — APPROVED** [review]

File: examples/wifi-echo/server/esp32/main/gen/endpoint_config.h (hunk 1 above)
Comment (rwalker-apple) [reply examples/wifi-echo/server/esp32/main/gen/endpoint_config.h:266]: you may want `&(generatedAttributes[84])` instead

Comment (bzbarsky-apple) [reply examples/wifi-echo/server/esp32/main/gen/endpoint_config.h:266]: `generatedAttributes` are the attribute descriptors in the `GENERATED_ATTRIBUTES` define.  So `generatedAttributes[18]` describes the attribute id, flags, number of bytes, etc, but all that is the same for both endpoints involved.

The storage location for an attribute is determined in `emAfReadOrWriteAttribute` by looping over endpoints.  Until the right one is found, we keep adding `emAfEndpoints[i].endpointType->endpointSize` to the offset, so after we examine endpoint 1 we will have advanced the offset to `84`.  Then when the right endpoint id is found, we loop over the clusters for the endpoint, increasing the offset by the total size of the cluster's attributes (from the relevant `GENERATED_CLUSTERS` entry) until the right cluster is found.  Then we loop over the attributes that cluster points to.

So if two endpoints have the same set of clusters they can point to the same entry in `GENERATED_CLUSTERS` and if two clusters have the same set of attributes (or if one is a subset of the other, with the same order), the two clusters can point to the same entry in `GENERATED_ATTRIBUTES`, afaict.  And other generated example apps do just that.

Comment (rwalker-apple) [reply examples/wifi-echo/server/esp32/main/gen/endpoint_config.h:266]: I think you've described how the storage code can reset itself to find the offset of the value in the storage array by assuming the data starts at the end of the previous endpoint's data.  Thanks, I think I get it.

This makes sense in general, but not in the case of singletons, whose data elements must point to the same spot in the data array.  Perhaps I don't understand what a "singleton" is in this environment.

@tecimovic can you help fix my feeble?

Comment (bzbarsky-apple) [reply examples/wifi-echo/server/esp32/main/gen/endpoint_config.h:266]: Singletons are treated as zero-sized for purposes of computing the space needed for an "endpoint's" or "cluster's" attributes, so they don't affect computation of other offsets.  And the location of a singleton is computed without using the built-up offset like so (in `emAfReadOrWriteAttribute` in `src/app/util/attribute-storage.c`):
```
                        EmberAfAttributeMetadata * am = &(cluster->attributes[attrIndex]);
...
                                uint8_t * attributeLocation =
                                    (am->mask & ATTRIBUTE_MASK_SINGLETON ? singletonAttributeLocation(am)
                                                                         : attributeData + attributeOffsetIndex);
```
That ends up indexing into `singletonAttributeData`, which is a totally separate array from `attributeData`....

Comment (rwalker-apple) [reply examples/wifi-echo/server/esp32/main/gen/endpoint_config.h:266]: I should have looked at the code, as you have.

Comment (rwalker-apple) [reply examples/wifi-echo/server/esp32/main/gen/endpoint_config.h:266]: do we want basic on every endpoint?

**Review by rwalker-apple — APPROVED** [review]

File: examples/wifi-echo/server/esp32/main/gen/endpoint_config.h (hunk 1 above)
Comment (bzbarsky-apple) [reply examples/wifi-echo/server/esp32/main/gen/endpoint_config.h:266]: I was torn on that.  For my purposes so far it's not needed, but I can pretty easily throw it in there....

**Review by kedars — APPROVED** [review]

**Review by mspang — APPROVED** [review]

**Pull request merged**